tools/wptrunner/wptrunner/executors/protocol.py

```diff
@@ -4,6 +4,24 @@
 from typing import ClassVar, List, Type
 
 
+def merge_dicts(target, source):
```

Comment:
What about something like
```py
def merge_dicts(target, source):
    if not (isinstance(target, dict) and isinstance(source, dict)):
        raise TypeError
    for (key, source_value) in source.items():
        if key not in target:
            target[key] = source_value
        else:
            if isinstance(source_value, dict) and isinstance(target[key], dict):
                merge_dicts(target[key], source_value)
            else:
                target[key] = source_value
   return rv
```
It means we can end up overwriting a dict value with a non-dict value which in context might be a programming error, but it's also simpler to understand what it does. Also I think the rollback you have has a bug; if there are two dict values in source and target, and the first merges successfully but the second doesn't we don't end up rolling back the changes to the first.

Reply:
Updated, and thanks for find the bug!